### ISSUE TITLE
High 'Total Blocking Time' In Lighthouse

### ISSUE THREAD

**Issue by 3200creative** [system-ui/theme-ui#980]
When using theme-ui I've noticed that the 'Total Blocking Time' is significantly increased. Primarily on mobile. This causes a lower Lighthouse score. I haven't experienced this using other styling methods (CSS, SASS)

Example: https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Ftheme-ui.com%2Fhome

I searched pretty extensively and haven't been able to locate any sites using theme-ui that don't experience this degradation in pagespeed.

**Comment by jxnblk** [comment]
I suspect this is due to the blocking script in the Gatsby plugin to set color mode before the page renders, but that shouldn't cause a significant issue. You can use Theme UI without the Gatsby plugin if you'd prefer to keep the FOUC behavior when using color modes in your site

**Comment by kenny-loveholidays** [comment]
We've noticed similar issues too. Especially since lighthouse v6 puts more emphasis on total blocking time.
See this https://github.com/preactjs/preact/issues/2531#issuecomment-627197933 for some experiments done on comparing `theme-ui` vs `emotion` only vs no `theme-ui` and no `emotion`

**Comment by 3200creative** [comment]
Thanks for the link! That further confirms what I'm experiencing. 

@jxnblk  What are your thoughts on this additional information?

**Comment by 3200creative** [comment]
This may help with the issue.

My 3200.pro site that I'm building over the next few weeks currently has a very simple homepage. This makes it great for testing because it's only got the basics. I was happy to see that it gets 100/100 in lighthouse. It's where I typically start on Gatsby websites.

**lighthouse:**

![google-pagespeed-100%](https://user-images.githubusercontent.com/3253469/87212153-aaabba80-c2e2-11ea-9724-377fe3574297.gif)

The same content in PageSpeed does significantly worse:

**In PageSpeed:**

![Screen Shot 2020-07-10 at 7 17 31 PM](https://user-images.githubusercontent.com/3253469/87212276-49d0b200-c2e3-11ea-8d40-9eef58883d19.jpg)

**Comment by MarkosKon** [comment]
I also have high TBT on a Gatsby site with `theme-ui`. 

After measuring the production build, I noticed that a good portion of React's hydrate (25%) was spend on calls to `get`, `css`, and `responsive` methods from [@theme-ui/css](https://github.com/system-ui/theme-ui/blob/master/packages/css/src/index.ts).   With x4 cpu slowdown that 25% is about 200-250 ms.

The same thing happens on a [side project site](https://character-table.netlify.app/). To measure, I use the _Performance_ tab of Chrome DevTools and the _bottom-up_ view. 

I wasn't able to observe something similar on gatsbyjs.com or the `theme-ui` docs site, so maybe I'm doing something wrong or overusing the `sx` prop.

**Comment by lachlanjc** [comment]
Related: #1234 will improve frontend performance

**Comment by tlgimenes** [comment]
Hi guys, it's been a while. I've managed to greatly reduced the blocking time by rewriting the theme-ui plugin and removing some features like dark mode. Also I had to hack emotion and I made a PR in emotion core in [here](https://github.com/emotion-js/emotion/pull/2159). 

The plugin is in [here](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-theme-ui). To try it just:
```yarn add @vtex/gatsby-plugin-theme-ui```
 
It's not well documented, however it should follow the same interface of the official plugin. If you guys happen to give it a try, please post you results so we can try merging these changes into the official gatsby plugin.

Thanks !

**Comment by lachlanjc** [comment]
Very cool! What did you discover? What would you suggest we change?

**Comment by AbdallahAbis** [comment]
@lachlanjc Any updates on this?

**Comment by lachlanjc** [comment]
@AbdallahAbis Nope! Would love contributions improving performance if you'd like to investigate.

**Comment by hasparus** [comment]
One problem is that the color modes detection has to be blocking — it's reading from localStorage, and we need it to happen ASAP to avoid color flash.

However, I know a few companies using Theme UI without color modes... I've advised people to manually use `@theme-ui/core` and add more packages as needed in such case. If this is more common than we expected we could reorganize the docs to deprioritize the umbrella package.

Regarding the Preact issue, I'm not sure if we can significantly limit the number of contexts we create without cutting some features.

cc @lachlanjc
